### ISSUE TITLE
SystemClock implementation is not monotonic

### ISSUE THREAD

**Issue by bachratyg** [aspnet/Security#874]
The SystemClock implementation has an observable discrepancy where a later call to SystemClock.UtcNow produces an earlier timestamp.
The current implementation states that
```C#
// the clock measures whole seconds only, to have integral expires_in results, and
// because milliseconds do not round-trip serialization formats
DateTimeOffset utcNow = DateTimeOffset.UtcNow; 
return utcNow.AddMilliseconds(-utcNow.Millisecond); 
```
however it will not result in whole seconds for e.g. new DateTimeOffset(636009776600935249, TimeSpan.Zero) will still have 0.0005249 fractional seconds after truncating milliseconds.
Assuming that the comment is right (e.g. the clock should measure integral seconds only) this should be the correct implementation:
```C#
return utcNow.AddTicks(-(utcNow.Ticks % TimeSpan.TicksPerSecond))
```

Not sure if any other repos have this issue. The ones i've found either do not expose their SystemClock implementation (e.g. KestrelHttpServer) or don't truncate milliseconds at all (e.g. Caching).

**Comment by Tratcher** [comment]
Did you find any behavioral side-effects from this? Or is it just a correctness issue?

**Comment by Eilon** [comment]
Closing this because we're not aware of anything that could be more than a fraction of a millisecond off.

**Comment by bachratyg** [comment]
I originally encountered this issue in the old owin implementation where the SystemClock seems to be positioned as a more generic component (opened an issue there too), then checked up on netcore to see if it's the same.
Apart from auth I use the clock to timestamp records created in the database. The client creates a resource and then queries all resources in a subsequent request (which is independent from the server pov). The current system time affects this latter query. We had no problem in unit tests (naturally, using a fixed clock), and our dev/test environment however in staging - due to the different network latency - this business function started returning inconsistent results for 1 in 3 requests with practically zero system load, marking the new resource as not yet valid. That was out of the blue since neither the client nor the database in the system provides any timestamps, that single shared SystemClock object handle time exclusively and since multiple network roundtrips are involved I did not expect the clock to move backwards. I can tolerate the occasional jump caused by os clock sync - and if not then I can configure it to keep the clock monotonic by clock rate corrections. This 1 in 3 however is beyond tolerable.
I did not find any issue with current security middlewares. However future middlewares may utilize fractional seconds. Imagine a theoretical security token that's not yet valid upon issue and triggers a reauth that issues a security token that's again not yet valid...etc. A clock change might trigger a single reauth but this continuously observed backstep may trigger an infinite auth loop if the timings are right.
So what I'm suggesting here is
1) fix the implementation to reflect the requirement in the comment by truncating sub-milliseconds. This is still a valid clock even if less precise than possible
2) fix the implementation by moving the truncation to the middleware that actually requires integral time and don't truncate in SystemClock at all. It's not the responsibility of ISystemClock to know about any serialization or roundtripping requirements.
3a) explicitly document that the middleware expects integral time and ignores microseconds - if it indeed depends on it and
3b) explicitly document that the SystemClock implementation performs this semi-truncation and microseconds should be disregarded as garbage.
4) make it internal

In short I did not find any side effects in current netcore - though I didn't look very thoroughly - and from this perspective it's more of a code correctness issue. However my own component depending on SystemClock misbehave with very high probability in certain environments.
________________________________

**Comment by Eilon** [comment]
@bachratyg ah, thanks for the extra info. In this case the SystemClock stuff is there just for testing purposes so that we can "mock" or "stub" the current time to create deterministic tests. Ideally it wouldn't even be exposed. We were investigating ways of avoiding having it show up, but weren't able to come up with a solution we liked in the time we have.

**Comment by Tratcher** [comment]
Yeah, this really is only intended for use in the cookie auth context where seconds are the lowest level of granularity. You're welcome to copy the abstraction, but the implementation is not well suited to your scenario.

Note that the implementation over in MemoryCache does not do such rounding: https://github.com/aspnet/Caching/blob/dev/src/Microsoft.Extensions.Caching.Abstractions/Internal/SystemClock.cs#L20

**Comment by bachratyg** [comment]
Thanks for the clarification.
One last question: are you aware of any harmful side effects if I plug in an implementation that does _not_ truncate? I'd like to use a single clock dependency/fixture through my app (possibly implementing ISystemClock from multiple packages as necessary) to simplify test code.

**Comment by Tratcher** [comment]
I don't remember anymore. I tried removing it and nothing failed, but I think most of the tests that would be affected use a TestSystemClock.
